File: jwt/src/main/java/com/member/jwt/jwt/CustomLogoutFilter.java
package com.member.jwt.jwt;

import com.member.jwt.repository.RefreshRepository;
import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.ServletRequest;
import jakarta.servlet.ServletResponse;
import jakarta.servlet.http.Cookie;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import org.springframework.web.filter.GenericFilterBean;

import java.io.IOException;

public class CustomLogoutFilter extends GenericFilterBean {

    private final JWTUtil jwtUtil;
    private final RefreshRepository refreshRepository;

    public CustomLogoutFilter(JWTUtil jwtUtil, RefreshRepository refreshRepository) {
        this.jwtUtil = jwtUtil;
        this.refreshRepository = refreshRepository;
    }

    @Override
    public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain) throws IOException, ServletException {
        doFilter((HttpServletRequest) request, (HttpServletResponse) response, chain);
    }

    private void doFilter(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws IOException, ServletException {
        System.out.println("Request URI: " + request.getRequestURI());
        System.out.println("Request Method: " + request.getMethod());
        System.out.println("Authorization Header: " + request.getHeader("Authorization"));

        // Logout path and method verification
        String requestURI = request.getRequestURI();
        if (!requestURI.equals("/logout") || !request.getMethod().equals("POST")) {
            filterChain.doFilter(request, response);
            return;
        }

        System.out.println("Logout filter activated for /logout");

        // Get refresh token from cookies
        String refresh = null;
        Cookie[] cookies = request.getCookies();
        if (cookies != null) {
            for (Cookie cookie : cookies) {
                if (cookie.getName().equals("refresh")) {
                    refresh = cookie.getValue();
                    break;
                }
            }
        }

        // Validate refresh token
        if (refresh == null || !jwtUtil.getCategory(refresh).equals("refresh") || !refreshRepository.existsByRefresh(refresh)) {
            response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
            response.getWriter().write("Invalid refresh token.");
            return;
        }

        // Remove refresh token from database
        refreshRepository.deleteByEmail(jwtUtil.getEmail(refresh));

        // Expire the refresh token cookie
        Cookie cookie = new Cookie("refresh", null);
        cookie.setMaxAge(0);
        cookie.setPath("/");

        response.addCookie(cookie);
        response.setStatus(HttpServletResponse.SC_OK);
        response.getWriter().write("Logged out successfully.");
    }
}

File: jwtKotlin/src/main/kotlin/com/member/jwtkotlin/jwt/CustomLogoutFilter.kt
package com.member.jwt.jwt

import com.member.jwt.repository.RefreshRepository
import jakarta.servlet.FilterChain
import jakarta.servlet.ServletException
import jakarta.servlet.ServletRequest
import jakarta.servlet.ServletResponse
import jakarta.servlet.http.Cookie
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.web.filter.GenericFilterBean
import java.io.IOException

class CustomLogoutFilter(jwtUtil: JWTUtil, private val refreshRepository: RefreshRepository) : GenericFilterBean() {
    private val jwtUtil: JWTUtil = jwtUtil

    @Throws(IOException::class, ServletException::class)
    override fun doFilter(request: ServletRequest, response: ServletResponse, chain: FilterChain) {
        doFilter(request as HttpServletRequest, response as HttpServletResponse, chain)
    }

    @Throws(IOException::class, ServletException::class)
    private fun doFilter(request: HttpServletRequest, response: HttpServletResponse, filterChain: FilterChain) {
        println("Request URI: " + request.requestURI)
        println("Request Method: " + request.method)
        println("Authorization Header: " + request.getHeader("Authorization"))

        // Logout path and method verification
        val requestURI = request.requestURI
        if (requestURI != "/logout" || request.method != "POST") {
            filterChain.doFilter(request, response)
            return
        }

        println("Logout filter activated for /logout")

        // Get refresh token from cookies
        var refresh: String? = null
        val cookies = request.cookies
        if (cookies != null) {
            for (cookie in cookies) {
                if (cookie.name == "refresh") {
                    refresh = cookie.value
                    break
                }
            }
        }

        // Validate refresh token
        if (refresh == null || !jwtUtil.getCategory(refresh).equals("refresh") || !refreshRepository.existsByRefresh(
                refresh
            )!!
        ) {
            response.status = HttpServletResponse.SC_BAD_REQUEST
            response.writer.write("Invalid refresh token.")
            return
        }

        // Remove refresh token from database
        refreshRepository.deleteByEmail(jwtUtil.getEmail(refresh))

        // Expire the refresh token cookie
        val cookie = Cookie("refresh", null)
        cookie.maxAge = 0
        cookie.path = "/"

        response.addCookie(cookie)
        response.status = HttpServletResponse.SC_OK
        response.writer.write("Logged out successfully.")
    }
}